File: dotnet/Net/Lan.cs
//----------------------------------------------------------------------------------------------------------------------------------------------------------------
//  Project :           Michitai.Lan
//  Author  :           Nichita Levandovici
//  Date    :           22.05.2026
//  Email   :           support@michitai.com
//  Website :           https://michitai.com
//----------------------------------------------------------------------------------------------------------------------------------------------------------------

using System;
using System.IO;
using System.Linq;
using System.Collections;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using System.Net;
using System.Net.Sockets;
using System.Net.Security;
using System.Net.NetworkInformation;
using System.Text;
using System.Runtime;
using System.Runtime.Serialization;


namespace Michitai.Lan.Net
{
    /// <summary>
    /// Provides static methods for LAN network operations and IP address retrieval.
    /// </summary>
    public static class Lan
    {
        /// <summary>
        /// Gets all local IPv4 addresses for the specified platform.
        /// </summary>
        /// <param name="platform">The platform to get addresses for.</param>
        /// <returns>IPAddress Array of all local NetworkInterfaces</returns>
        public static IPAddress[] LocalIPv4Addresses(EPlatform platform)
    {
        var ip_addresses = new List<IPAddress>();

        int count = 0;


        if ((platform & EPlatform.Windows) == EPlatform.Windows ||
            (platform & EPlatform.Linux) == EPlatform.Linux ||
            (platform & EPlatform.MacOS) == EPlatform.MacOS)
        {
            var ni = NetworkInterface.GetAllNetworkInterfaces();

            foreach (NetworkInterface item in ni)
            {
                if (item.OperationalStatus == OperationalStatus.Up)
                {
                    foreach (UnicastIPAddressInformation ip in item.GetIPProperties().UnicastAddresses)
                    {
                        if (ip.Address.AddressFamily == AddressFamily.InterNetwork & !IPAddress.IsLoopback(ip.Address))
                        {
                            ip_addresses.Add(ip.Address);

                            count++;
                        }
                    }
                }
            }
        }
        else if ((platform & EPlatform.Android) == EPlatform.Android ||
                (platform & EPlatform.IOS) == EPlatform.IOS)
        {
            var host = Dns.GetHostEntry(Dns.GetHostName());

            foreach (var ip in host.AddressList)
            {
                if (ip.AddressFamily == AddressFamily.InterNetwork && !IPAddress.IsLoopback(ip))
                {
                    ip_addresses.Add(ip);

                    count++;
                }
            }
        }


        if (count > 0)
        {
            return ip_addresses.ToArray();
        }
        else
        {
            return new IPAddress[0];
        }
    }

        /// <summary>
        /// Gets all local IPv4 broadcast masks for the specified platform.
        /// </summary>
        /// <param name="platform">The platform to get masks for.</param>
        /// <returns>IPAddress Masks Array of all local NetworkInterfaces. Example 192.168.0.255</returns>
        public static IPAddress[] LocalIPv4Masks(EPlatform platform)
    {
        var ips = LocalIPv4Addresses(platform);

        return LocalIPv4Masks(ips);
    }

        /// <summary>
        /// Converts IPv4 addresses to broadcast masks.
        /// </summary>
        /// <param name="iPv4Addresses">IPAddresses Array. Example 192.168.0.1</param>
        /// <returns>IPAddresses Masks Array. Example 192.168.0.255</returns>
        public static IPAddress[] LocalIPv4Masks(IPAddress[] iPv4Addresses)
    {
        return iPv4Addresses.Select(ip =>
        {
            var bytes = ip.GetAddressBytes();

            bytes[bytes.Length - 1] = 255;

            return new IPAddress(bytes);

        }).ToArray();
    }



        /// <summary>
        /// Attempts to get all local IPv4 addresses for the specified platform.
        /// </summary>
        /// <param name="platform">The platform to get addresses for.</param>
        /// <param name="ipAddresses">When this method returns, contains the IP addresses if successful.</param>
        /// <returns>True if addresses were found; otherwise, false.</returns>
        public static bool TryGetLocalIPv4Addresses(EPlatform platform, out IPAddress[] ipAddresses)
    {
        ipAddresses = LocalIPv4Addresses(platform);

        return ipAddresses.Length > 0;
    }

        /// <summary>
        /// Attempts to get all local IPv4 broadcast masks for the specified platform.
        /// </summary>
        /// <param name="platform">The platform to get masks for.</param>
        /// <param name="masks">When this method returns, contains the masks if successful.</param>
        /// <returns>True if masks were found; otherwise, false.</returns>
        public static bool TryGetLocalIPv4Masks(EPlatform platform, out IPAddress[] masks)
    {
        masks = LocalIPv4Masks(platform);

        return masks.Length > 0;
    }

        /// <summary>
        /// Attempts to get all local IPv4 broadcast masks as strings for the specified platform.
        /// </summary>
        /// <param name="platform">The platform to get masks for.</param>
        /// <param name="masks">When this method returns, contains the mask strings if successful.</param>
        /// <returns>True if masks were found; otherwise, false.</returns>
        public static bool TryGetLocalIPv4Masks(EPlatform platform, out string[] masks)
    {
        masks = LocalIPv4Masks(platform).Select(ip => ip.ToString()).ToArray();

        return masks.Length > 0;
    }
}
}

File: cpp/src/Net/Lan.cpp
//----------------------------------------------------------------------------------------------------------------------------------------------------------------
//  Project :           Michitai.Lan
//  Author  :           Nichita Levandovici
//  Date    :           22.05.2026
//  Email   :           support@michitai.com
//  Website :           https://michitai.com
//----------------------------------------------------------------------------------------------------------------------------------------------------------------

#include <Michitai/Lan/Net/Lan.hpp>
#include "SocketInternal.hpp"

#ifdef _WIN32
    #include <iphlpapi.h>
#else
    #include <ifaddrs.h>
    #include <net/if.h>
    #include <netdb.h>
#endif


namespace Michitai::Lan::Net
{
    std::vector<IPAddress> Lan::LocalIPv4Addresses(EPlatform platform)
    {
        std::vector<IPAddress> addresses;

        if (HasFlag(platform, EPlatform::Windows) ||
            HasFlag(platform, EPlatform::Linux) ||
            HasFlag(platform, EPlatform::MacOS))
        {
            Detail::EnsureSocketsInitialized();

#ifdef _WIN32
            ULONG size = 0;
            if (GetAdaptersAddresses(AF_INET, GAA_FLAG_SKIP_ANYCAST | GAA_FLAG_SKIP_MULTICAST |
                                     GAA_FLAG_SKIP_DNS_SERVER, nullptr, nullptr, &size) != ERROR_BUFFER_OVERFLOW)
            {
                return addresses;
            }

            std::vector<std::uint8_t> buffer(size);
            auto* adapters = reinterpret_cast<IP_ADAPTER_ADDRESSES*>(buffer.data());

            if (GetAdaptersAddresses(AF_INET, GAA_FLAG_SKIP_ANYCAST | GAA_FLAG_SKIP_MULTICAST |
                                     GAA_FLAG_SKIP_DNS_SERVER, nullptr, adapters, &size) != NO_ERROR)
            {
                return addresses;
            }

            for (IP_ADAPTER_ADDRESSES* adapter = adapters; adapter; adapter = adapter->Next)
            {
                if (adapter->OperStatus != IfOperStatusUp)
                    continue;

                for (IP_ADAPTER_UNICAST_ADDRESS* unicast = adapter->FirstUnicastAddress;
                     unicast; unicast = unicast->Next)
                {
                    if (unicast->Address.lpSockaddr->sa_family != AF_INET)
                        continue;

                    auto* ipv4 = reinterpret_cast<sockaddr_in*>(unicast->Address.lpSockaddr);
                    IPAddress address = IPAddress::FromValue(ntohl(ipv4->sin_addr.s_addr));

                    if (!IPAddress::IsLoopback(address))
                        addresses.push_back(address);
                }
            }
#else
            ifaddrs* list = nullptr;
            if (getifaddrs(&list) != 0)
                return addresses;

            for (ifaddrs* item = list; item; item = item->ifa_next)
            {
                if (!item->ifa_addr || item->ifa_addr->sa_family != AF_INET)
                    continue;
                if (!(item->ifa_flags & IFF_UP))
                    continue;

                auto* ipv4 = reinterpret_cast<sockaddr_in*>(item->ifa_addr);
                IPAddress address = IPAddress::FromValue(ntohl(ipv4->sin_addr.s_addr));

                if (!IPAddress::IsLoopback(address))
                    addresses.push_back(address);
            }

            freeifaddrs(list);
#endif
        }
        else if (HasFlag(platform, EPlatform::Android) || HasFlag(platform, EPlatform::IOS))
        {
            Detail::EnsureSocketsInitialized();

            char host[256] = {};
            if (::gethostname(host, sizeof(host) - 1) != 0)
                return addresses;

            addrinfo hints{};
            hints.ai_family = AF_INET;

            addrinfo* result = nullptr;
            if (::getaddrinfo(host, nullptr, &hints, &result) != 0 || !result)
                return addresses;

            for (addrinfo* item = result; item; item = item->ai_next)
            {
                if (item->ai_family != AF_INET)
                    continue;

                auto* ipv4 = reinterpret_cast<sockaddr_in*>(item->ai_addr);
                IPAddress address = IPAddress::FromValue(ntohl(ipv4->sin_addr.s_addr));

                if (!IPAddress::IsLoopback(address))
                    addresses.push_back(address);
            }

            ::freeaddrinfo(result);
        }

        return addresses;
    }

    std::vector<IPAddress> Lan::LocalIPv4Masks(EPlatform platform)
    {
        return LocalIPv4Masks(LocalIPv4Addresses(platform));
    }

    std::vector<IPAddress> Lan::LocalIPv4Masks(const std::vector<IPAddress>& addresses)
    {
        std::vector<IPAddress> masks;
        masks.reserve(addresses.size());

        for (const IPAddress& address : addresses)
        {
            masks.push_back(IPAddress::FromValue(address.Value() | 0xFF));
        }

        return masks;
    }
}
